pkg/innerring/contracts.go

```diff
@@ -121,11 +145,60 @@ func parseAlphabetContracts(cfg *viper.Viper, morph *client.Client) (alphabetCon
 	return alpha, nil
 }
 
-func parseContract(cfg *viper.Viper, morph *client.Client, cfgName, nnsName string) (res util.Uint160, err error) {
-	contractStr := cfg.GetString(cfgName)
-	if len(contractStr) == 0 {
-		return morph.NNSContractAddress(nnsName)
+type nnsContext struct {
+	context.Context
+
+	nnsDeployed bool
+}
+
+func parseContract(ctx *nnsContext, _logger *logger.Logger, cfg *viper.Viper, morph *client.Client, cfgName, nnsName string) (res util.Uint160, err error) {
+	if cfg.IsSet(cfgName) {
+		_logger.Debug("decoding configured contract...", zap.String("config key", cfgName))
+		return util.Uint160DecodeStringLE(cfg.GetString(cfgName))
+	}
+
+	const pollInterval = 5 * time.Second // no too long and not so frequently
+
+	if !ctx.nnsDeployed {
+		for {
+			_logger.Info("waiting for NNS contract to be deployed...")
+
+			select {
+			case <-ctx.Done():
+				return res, fmt.Errorf("waiting for NNS contract: %w", ctx.Err())
+			default:
+			}
+
+			_, err = morph.NNSHash()
+			if err == nil {
+				ctx.nnsDeployed = true
+				break
+			}
+
+			// FIXME: unstable approach, but currently there is no other way
+			//  Track neofs-node#2285
+			if !strings.Contains(err.Error(), "Unknown contract") {
+				return
+			}
+
+			time.Sleep(pollInterval)
+		}
 	}
 
-	return util.Uint160DecodeStringLE(contractStr)
+	for {
+		_logger.Info("waiting for contract registration in the NNS...", zap.String("name", nnsName))
```

Comment:
Just a question: NNS contract address is being regirtered in the NNS contract itself, right?

Reply:
if I understand the question correctly: we have an unspoken prerequisite for deploying the NeoFS NNS contract with ID=1

Reply:
@AnnaShaleva : this is one is about non-NNS contracts registering in NNS as a part of their deployment procedure.

Reply:
Oh, I got it, `nnsName` is not the "Neo name service`, it's just the method's parameter and the name of some other neofs contract.